Add an option, third_party_invite_username_separator_string, that allows redacting separate portions of a third party invite address

anoadragon453 opened the pull request:
When generating a third-party invite, Sydent tries to obfuscate part of the invited email address before creating and sending off an event that will eventually be inserted into the relevant room. This address is obfuscated so that it can both:

* Be recognised by whomever created it and additionally distinguished from other third-party invites by all users.
* Hide the full address in order to prevent leaking private information.

The username and domain in an email address are obfuscated separately. However, this can pose an issue when emails look something like "John.Smith@example.com". This could be obfuscated to "John.Sm...@exam...".

While this has hidden the user's last name, their first name is still visible!

This PR adds an option that lets you configure a separator for email address usernames. The resulting strings after splitting the username using this separator will be obfuscated individually. So, if this option were set to `.`, then "John.Smith@example.com" would become, say, "Jo...Sm...@exam...".

Note that there currently only exists an option for usernames, as that's all that was requested.

clokep commented on sydent/http/servlets/store_invite_servlet.py:
```diff
@@ -147,8 +147,26 @@ def redact_email_address(self, address):
+        username_components = username.split(separator) if separator else [username]
+
+        # Obfuscate each individual component
+        redacted_components = [
+            self._redact(component, self.sydent.username_obfuscate_characters)
+            for component in username_components if component != ""
+        ]
```
I don't know if it makes sense to drop the empty portions, if you have `donald--fauntleroy--duck@disney.com` I think I'd expect this (with a count of 3) to become `don...--fau...--duc...`, not `don...-fau...-duc`.

I think this also guarantees that `redacted_components` will be truthy also?

anoadragon453 replied:
> I think this also guarantees that redacted_components will be truthy also?

Well, if the username == the separator then `redacted_components` would be an empty list, and thus `False`.

Though I'm likely going to remove `if component != ""` based on other comments.

anoadragon453 replied:
Well, removing `if component != ""` has the side-effect of turning:

`applejack-@someexample.com` into `applej...-...@someexample.com`

Which may be a fine compromise?

clokep commented on sydent/sydent.py:
```diff
@@ -174,6 +174,18 @@ def set_from_comma_sep_string(rawstr):
+        # A string to separate multiple components of the username portion of an email address.
+        # For instance, if "." is set, then "alice.smith@example.com" would result
+        # in both "alice" and "smith" being individually obfuscated. Resulting in
+        # "ali...smi...@example.com" for example.
```
This example seems wrong, wouldn't it end up being `ali....smi...@example.com`, note the four dots after `ali` (you end up with `...` from the redaction and then one from the separator).

It seems like it would have been nicer to use the ellipses character (`…`) instead of 3 dots (`...`) here.

anoadragon453 replied:
Oh, using `…` is an excellent idea! Yes, with the ` if component != ""` check I was trying to eliminate the edge case of having four dots, but I think this might be a more elegant solution.

Of course doing so is going to conflict with each of the other PRs, but the change itself won't have any backwards compatibility concerns.

I might leave the edge case of four dots in this PR, and make another one when all 3 PRs are merged to replace `...` with the ellipses.

anoadragon453 replied:
I've also changed the example to use `-`, which is clearer.

clokep commented on tests/test_invites.py:
```diff
@@ -90,6 +90,27 @@ def test_invited_email_address_obfuscation(self):
+        email_address = "johnathon-jingle-smithington@company-town.notarealtld"
+        redacted_address = store_invite_servlet.redact_email_address(email_address)
+        # Each individual component of the username should be obfuscated, but not the domain
+        self.assertEqual(redacted_address, "johnat...-jin...-smithi...@company-...")
```
Shouldn't this be `@compan...`? This seems to show it also splitting the domain on `-`?

anoadragon453 replied:
`email.third_party_invite_domain_obfuscate_characters` is set to 8 for this test, so it should split there. But you're right, splitting on `-` is quite confusing.

clokep replied:
Ah, it just happens to be that that's 8 characters. 🤦 Maybe we should just change the domain count for this test so that it doesn't seem to be doing the wrong thing.

clokep commented on tests/test_invites.py:
```diff
@@ -90,6 +90,27 @@ def test_invited_email_address_obfuscation(self):
+        self.assertEqual(redacted_address, "johnat...-jin...-smithi...@company-...")
+
+        # Try one with a separator at a word boundary
+        self.sydent.third_party_invite_username_separator_string = "."
```
Not sure why you need to change the separator here? The tests would probably be clearer using `-` since `...` is the replacement.

anoadragon453 replied:
This was intended to check the four dots issue, but as we'll be solving that in another way this test can be changed to use `-` instead.

clokep commented on tests/test_invites.py:
```diff
@@ -90,6 +90,27 @@ def test_invited_email_address_obfuscation(self):
+        redacted_address = store_invite_servlet.redact_email_address(email_address)
+        self.assertEqual(redacted_address, "applej...@someexam...")
+
+        # Try one where a separator is the username
```
```suggestion
        # Try one where the username is just the separator.
```

clokep commented on tests/test_invites.py:
```diff
@@ -90,6 +90,27 @@ def test_invited_email_address_obfuscation(self):
+
+        email_address = ".@someexample.com"
+        redacted_address = store_invite_servlet.redact_email_address(email_address)
+        self.assertEqual(redacted_address, ".@someexam...")
```
It would be nice to add a test with a double separator (`--`) and what the expectation is there.

clokep reviewed (dismissed):
Seems OK, but I think there's an edge-case missing.

anoadragon453 commented:
Right, lots of changes. I replaced the separate redaction code with something a little more involved, but one that I believe covers all of the edge cases. I've updated the tests accordingly and based on your feedback of what one might be more likely to expect.

clokep commented on sydent/http/servlets/store_invite_servlet.py:
```diff
@@ -147,8 +147,34 @@ def redact_email_address(self, address):
+            # username individually.
+            redacted_username = ""
+
+            username_components = username.split(separator)
+            for index, component in enumerate(username_components):
+                if component:
+                    # Redact this component and append it to the final string
+                    redacted_username += (
+                        self._redact(component, self.sydent.username_obfuscate_characters)
+                    )
+
+                # Append the separator separately. This handles the case of multiple
+                # separators appearing sequentially in the username.
+                #
+                # The conditional here is to prevent a separator being appended to the end.
+                if index != len(username_components) - 1:
+                    redacted_username += separator
```
I think this can be simplified, I'm not 100% sure it is identical, but I think it is.

```suggestion
            # Redact each component individually, if it has content. (No content implies multiple sequential separators.)
            redacted_components = [
                self._redact(component, self.sydent.username_obfuscate_characters) if component else "" for component in username.split(separator)
            ]
            redacted_username = separator.join(username_components)
```

anoadragon453 replied:
Hah, yep this works. And to think I was so close :) Thanks!

clokep commented on tests/test_invites.py:
```diff
@@ -90,6 +90,30 @@ def test_invited_email_address_obfuscation(self):
 
+        # Try using a username separator string
+        self.sydent.third_party_invite_username_separator_string = "-"
+        email_address = "johnathon-jingle-smithington@smiths.notarealtld"
+        redacted_address = store_invite_servlet.redact_email_address(email_address)
+        # Each individual component of the username should be obfuscated, but not the domain
+        self.assertEqual(redacted_address, "johnat...-jin...-smithi...@smiths.n...")
```
I think the following makes it clearer that this does *not* apply to the domain:

```suggestion
        email_address = "johnathon-jingle-smithington@john-smith.notarealtld"
        redacted_address = store_invite_servlet.redact_email_address(email_address)
        # Each individual component of the username should be obfuscated, but not the domain
        self.assertEqual(redacted_address, "johnat...-jin...-smithi...@john-smi...")
```

clokep reviewed (changes requested):
Looks better, but I think the code can be simplified a bit!

anoadragon453 commented on sydent/http/servlets/store_invite_servlet.py:
```diff
@@ -147,8 +147,34 @@ def redact_email_address(self, address):
 
-        # Obfuscate strings
-        redacted_username = self._redact(username, self.sydent.username_obfuscate_characters)
+        # Obfuscate the username portion
+
```
Err, I think a change might've been accidentally lost here :)

clokep replied:
It removes the blank line after the comment...GitHub doesn't seem to show that well. 😄

anoadragon453 replied:
Ahh, gotcha! Yes, that does make sense now that everything is encapsulated in an `if`.

clokep reviewed (approved):
LGTM!

giomfo reviewed (approved):
LGTM
